File: src/main/java/com/portofino/realtrainmodunofficial/vehicle/VehicleRegistry.java
package com.portofino.realtrainmodunofficial.vehicle;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Locale;

public final class VehicleRegistry {
    private static final List<VehicleDefinition> DEFINITIONS = new ArrayList<>();
    private static final Map<String, VehicleDefinition> BY_ID = new HashMap<>();
    private static int selectedIndex = 0;

    private VehicleRegistry() {
    }

    public static void setDefinitions(List<VehicleDefinition> defs) {
        DEFINITIONS.clear();
        BY_ID.clear();
        Map<String, VehicleDefinition> unique = new LinkedHashMap<>();
        for (VehicleDefinition d : defs) {
            if (d == null) {
                continue;
            }
            unique.putIfAbsent(dedupeKey(d), d);
        }
        for (VehicleDefinition d : unique.values()) {
            DEFINITIONS.add(d);
            BY_ID.putIfAbsent(d.getId(), d);
        }
        if (selectedIndex >= DEFINITIONS.size() || isHiddenDefault(getSelected())) {
            selectedIndex = firstUsableIndex();
        }
    }

    public static List<VehicleDefinition> getAll() {
        return List.copyOf(DEFINITIONS);
    }

    public static VehicleDefinition getById(String id) {
        return id == null ? null : BY_ID.get(id);
    }

    public static VehicleDefinition getSelected() {
        if (DEFINITIONS.isEmpty()) return null;
        if (selectedIndex < 0 || selectedIndex >= DEFINITIONS.size()) selectedIndex = firstUsableIndex();
        return DEFINITIONS.get(selectedIndex);
    }

    public static void setSelectedIndex(int i) {
        if (i >= 0 && i < DEFINITIONS.size()) selectedIndex = i;
    }

    private static int firstUsableIndex() {
        for (int i = 0; i < DEFINITIONS.size(); i++) {
            if (!isHiddenDefault(DEFINITIONS.get(i))) {
                return i;
            }
        }
        return 0;
    }

    private static boolean isHiddenDefault(VehicleDefinition definition) {
        if (definition == null) {
            return false;
        }
        String packName = safe(definition.getPackName());
        String id = safe(definition.getId());
        String displayName = safe(definition.getDisplayName());
        return "basic_train".equalsIgnoreCase(packName)
            || id.toLowerCase().contains("basic_train")
            || displayName.toLowerCase().contains("basic_train");
    }

    private static String safe(String value) {
        return value == null ? "" : value;
    }

    private static String dedupeKey(VehicleDefinition definition) {
        String display = safe(definition.getDisplayName());
        String button = safe(definition.getButtonTexture());
        String model = safe(definition.getModelFile());
        if (!display.isBlank() && !button.isBlank()) {
            return (display + "|" + button).toLowerCase(Locale.ROOT);
        }
        return (display + "|" + model).toLowerCase(Locale.ROOT);
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/vehicle/VehicleRegistry.kt
package cc.mirukuneko.realtrainmodrenewed.vehicle

import java.util.Locale

object VehicleRegistry {
    private val definitions: MutableList<VehicleDefinition> = ArrayList()
    private val byId: MutableMap<String, VehicleDefinition> = HashMap()
    private val idCounts: MutableMap<String, Int> = HashMap()
    private var selectedIndex = 0

    @JvmStatic
    fun setDefinitions(defs: List<VehicleDefinition?>) {
        definitions.clear()
        byId.clear()
        idCounts.clear()
        val unique = LinkedHashMap<String, VehicleDefinition>()
        for (definition in defs) {
            if (definition != null) {
                unique.putIfAbsent(dedupeKey(definition), definition)
            }
        }
        for (definition in unique.values) {
            idCounts[definition.id] = (idCounts[definition.id] ?: 0) + 1
        }
        for (definition in unique.values) {
            definitions.add(definition)
            byId.putIfAbsent(definition.id, definition)
            byId[packScopedId(definition)] = definition
        }
        if (selectedIndex >= definitions.size || isHiddenDefault(getSelected())) {
            selectedIndex = firstUsableIndex()
        }
    }

    @JvmStatic
    fun getAll(): List<VehicleDefinition> = definitions.toList()

    @JvmStatic
    fun getById(id: String?): VehicleDefinition? = if (id == null) null else byId[id]

    @JvmStatic
    fun getSelectionId(definition: VehicleDefinition): String =
        if ((idCounts[definition.id] ?: 0) > 1) packScopedId(definition) else definition.id

    @JvmStatic
    fun getSelected(): VehicleDefinition? {
        if (definitions.isEmpty()) {
            return null
        }
        if (selectedIndex !in definitions.indices) {
            selectedIndex = firstUsableIndex()
        }
        return definitions[selectedIndex]
    }

    @JvmStatic
    fun setSelectedIndex(i: Int) {
        if (i in definitions.indices) {
            selectedIndex = i
        }
    }

    private fun firstUsableIndex(): Int {
        for (i in definitions.indices) {
            if (!isHiddenDefault(definitions[i])) {
                return i
            }
        }
        return 0
    }

    private fun isHiddenDefault(definition: VehicleDefinition?): Boolean {
        if (definition == null) {
            return false
        }
        val packName = safe(definition.packName)
        val id = safe(definition.id)
        val displayName = safe(definition.displayName)
        return packName.equals("basic_train", ignoreCase = true) ||
            id.lowercase(Locale.ROOT).contains("basic_train") ||
            displayName.lowercase(Locale.ROOT).contains("basic_train")
    }

    private fun safe(value: String?): String = value ?: ""

    private fun dedupeKey(definition: VehicleDefinition): String {
        val id = safe(definition.id)
        val pack = safe(definition.packName)
        val model = safe(definition.modelFile)
        return "$pack|$id|$model".lowercase(Locale.ROOT)
    }

    private fun packScopedId(definition: VehicleDefinition): String =
        "${definition.packName}:${definition.id}"
}
